relay-server/src/services/projects/cache/service.rs

```diff
@@ -0,0 +1,221 @@
+    }
+
+    /// Consumes and starts a [`ProjectCacheService`].
+    ///
+    /// Returns a [`ProjectCacheHandle`] to access the cache concurrently.
+    pub fn start(self) -> ProjectCacheHandle {
+        let (addr, addr_rx) = relay_system::channel(Self::name());
+
+        let handle = ProjectCacheHandle {
+            shared: self.store.shared(),
+            config: Arc::clone(&self.config),
+            service: addr,
+            project_events: self.project_events_tx.clone(),
+        };
+
+        self.spawn_handler(addr_rx);
+
+        handle
+    }
+
+    /// Schedules a new [`Fetch`] and delivers the result to the [`Self::project_update_tx`] channel.
+    fn schedule_fetch(&self, fetch: Fetch) {
+        let source = self.source.clone();
+        let project_updates = self.project_update_tx.clone();
+
+        tokio::spawn(async move {
+            tokio::time::sleep_until(fetch.when()).await;
+
+            // TODO: cached state for delta fetches, maybe this should just be a revision?
+            let state = match source
+                .fetch(fetch.project_key(), false, fetch.revision())
+                .await
+            {
+                Ok(result) => result,
+                Err(err) => {
+                    relay_log::error!(
+                        tags.project_key = fetch.project_key().as_str(),
+                        error = &err as &dyn std::error::Error,
+                        "failed to fetch project from source: {fetch:?}"
+                    );
+
+                    // TODO: change this to ProjectState::Pending once we consider it safe to do so.
+                    // see https://github.com/getsentry/relay/pull/4140.
+                    ProjectState::Disabled.into()
+                }
+            };
+
+            let _ = project_updates.send(fetch.complete(state));
+        });
+    }
+}
+
+/// All [`ProjectCacheService`] message handlers.
+impl ProjectCacheService {
+    fn handle_fetch(&mut self, project_key: ProjectKey) {
+        if let Some(fetch) = self.store.try_begin_fetch(project_key, &self.config) {
+            self.schedule_fetch(fetch);
+        }
+    }
+
+    fn handle_project_update(&mut self, fetch: CompletedFetch) {
+        let project_key = fetch.project_key();
+
+        if let Some(fetch) = self.store.complete_fetch(fetch, &self.config) {
+            relay_log::trace!(
+                project_key = fetch.project_key().as_str(),
+                "re-scheduling project fetch: {fetch:?}"
+            );
+            self.schedule_fetch(fetch);
+            return;
+        }
+
+        let _ = self
+            .project_events_tx
+            .send(ProjectEvent::Ready(project_key));
+    }
+
+    fn handle_evict_stale_projects(&mut self) {
+        let on_evict = |project_key| {
+            let _ = self
+                .project_events_tx
+                .send(ProjectEvent::Evicted(project_key));
+        };
+
+        self.store.evict_stale_projects(&self.config, on_evict);
```

Comment:
You mentioned in the PR description that there is less garbage in this version, why is that? Isn't the amount of project configs that need deconstruction still the same?

Reply:
It does a slightly better handling of revisions which re-creates less projects. Where we before did always send back a project from the source, the source itself can now signal that there was no change. This is relevant for the sources where we peek into the deserialized contents and compare revisions (Redis).